src/media/local-track.ts

```diff
@@ -52,77 +34,52 @@ export type TrackEffect = any;
-
-  /**
-   * The MediaStream that represents the current 'output' of this track.
-   * If no effects have been added, this will be the original MediaStream;
-   * if effects have been added, this will represent the output of the 'last'
-   * effect in the pipeline.
-   */
-  private _underlyingStream!: MediaStream;
-
-  /**
-   * The MediaStream that was passed to the constructor.
-   */
-  private originalStream: MediaStream;
-
+export abstract class LocalTrack extends Track {
   private isPublished = false;
 
   private effects: Map<string, TrackEffect> = new Map();
 
-  label: string;
-
   /**
-   * Constructor for the Track class. Creates an empty CoreLocalTrack or uses an existing one.
+   * Get an effect by name.
    *
-   * @param stream - The MediaStream for this LocalTrack.
+   * @param name - The effect name.
+   * @returns A MicrophoneEffect.
    */
-  constructor(stream: MediaStream) {
-    super();
-    this.originalStream = stream;
-    this.underlyingStream = stream;
-    // Effects create a new label but we want to retain the original one.
-    this.label = this.underlyingStream.getTracks()[0].label;
+  getEffect(name: string): TrackEffect | undefined {
+    const effect = this.effects.get(name);
 
-    /**
-     * Emit ended event when the underlying track ends.
-     */
-    this.underlyingTrack.onended = () => {
-      this.emit(LocalTrackEvents.Ended, { trackState: this.trackState });
-    };
-  }
+    if (!effect) {
+      logger.log(`No effect found with name '${name}'`);
+    }
 
-  /**
-   * Get the kind of this track.
-   * NOTE(brian): the need for this will likely go away once we get the rest of the track hierarchy
-   * done, as we can use something like instanceof instead.
-   *
-   * @returns - The kind of this track, as a MediaStreamTrackKind.
-   */
-  get kind(): MediaStreamTrackKind {
-    return this.underlyingTrack.kind as MediaStreamTrackKind;
+    return effect;
   }
 
   /**
-   * Get id of this track.
+   * Adds an effect to a local track.
    *
-   * @returns The id of this track.
+   * @param name - The name of the effect.
+   * @param effect - The effect to add.
    */
-  get id(): string {
-    return this.underlyingTrack.id;
+  async addEffect(name: string, effect: TrackEffect): Promise<void> {
+    await effect.load(new MediaStream([this.getMediaStreamTrack()]));
+    this.setMediaStreamTrackWithEffects(effect.getUnderlyingStream().getAudioTracks()[0]);
+    this.effects.set(name, effect);
   }
 
   /**
-   * Get muted state of this track.
-   *
-   * @returns The muted state of this track.
+   * Disposes the effect already applied .
    */
-  get muted(): boolean {
-    return !this.underlyingTrack.enabled;
+  disposeEffects(): void {
+    if (this.effects.size > 0) {
+      this.effects.forEach((effect: TrackEffect) => effect.dispose());
+      this.effects.clear();
+
+      // this.underlyingStream = this.originalStream;
+      // this.emit(LocalTrackEvents.UnderlyingTrackChange);
```

Comment:
Are these needed? Should either restore them or delete the lines

Reply:
removed the comment 

i still need to attach events for the effect track for onended and onmuted 

when applying effects i have to save  both the local track and track with effect in one place , because both are in track class the event for track change would first from there 

i dont see any harm firing the event form track class as localTrack also extends the track if it want it can also fire change event